(#1276) List command with last updated date

swaran18999 opened the pull request:
## Description Of Changes
- Add --show-last-updated-date option to choco list to display the last date a package was installed/updated, formatted as yyyy-MM-dd HH:mm:ss. Falls back to “Last updated not available” when metadata is missing.
- Add --order-by-last-updated-date option to choco list to order results by the package’s last installed/updated date by mapping to PackageOrder.LastPublished.
- Update PowerShell tab completion (src/chocolatey.resources/helpers/ChocolateyTabExpansion.ps1) to include both new switches in the list command’s completion options.

## Motivation and Context
- Administrators and users often need quick visibility into when packages were last installed or updated, and to sort lists by recency for auditing and maintenance workflows. Previously, choco list did not expose this information nor allow sorting by it.
- These additions improve usability by displaying last-change timestamps inline with list results.
- This work addresses the needs discussed in issue #1276

## Testing
- Flag parsing for --show-last-updated-date sets ShowLastUpdatedDate = true; defaults to false when omitted. Output includes formatted date yyyy-MM-dd HH:mm:ss when Published exists.
- Sorting places packages in the expected order by last updated date. 
- Correct order when --order-by-last-updated-date is used.
- Behavior is consistent with --id-only/--by-id-only (date not appended, sorting still applied).

### Operating Systems Testing
- Windows 10

## Change Types Made
* [ ] Bug fix (non-breaking change).
* [x] Feature / Enhancement (non-breaking change).
* [ ] Breaking change (fix or feature that could cause existing functionality to change).
* [ ] Documentation changes.
* [ ] PowerShell code changes.

## Change Checklist
* [x] Requires a change to the documentation.
* [ ] Documentation has been updated.
* [ ] Tests to cover my changes, have been added.
* [x] All new and existing tests passed?
* [ ] PowerShell code changes: PowerShell v3 compatibility checked?

## Related Issue
Partially fixes #1276

swaran18999 committed:
(#1276) Add option to show last updated date in `choco list`

Add `--show-last-updated-date` option to the `list` command. When
enabled, this displays the last date the package was installed or
updated, alongside the usual metadata fields.

Previously, users had no clear indication of when a local package was
last changed. This new option provides better visibility and helps
with package management, especially when filtering or sorting package
lists by recency.

Implementation details:
- Introduced new switch `show-last-updated-date`.
- Logs formatted date via `package.Published` (or fallback string if
  unavailable).
- Integrated into existing `list` command output formatting.

With this change, users can more easily audit their package states
and track updates over time.

swaran18999 committed:
(#1276) Add option to show last updated date in `choco list`

Add `--show-last-updated-date` option to the `list` command. When
enabled, this displays the last date the package was installed or
updated, alongside the usual metadata fields.

Previously, users had no clear indication of when a local package was
last changed. This new option provides better visibility and helps
with package management, especially when filtering or sorting package
lists by recency.

Implementation details:
- Introduced new switch `show-last-updated-date`.
- Logs formatted date via `package.Published` (or fallback string if
  unavailable).
- Integrated into existing `list` command output formatting.

With this change, users can more easily audit their package states
and track updates over time.

swaran18999 committed:
(#1276) Add option to order packages by last updated date

Add `--order-by-last-updated-date` option to the `list` command. When
used, the command output is ordered by the date each package was last
installed or updated.

Previously, package listings were only ordered by name or version.
This limited users who wanted to quickly identify the most recently
changed packages. With this new option, users gain more control over
list output and can better manage their environments based on package
activity.

Implementation details:
- Introduced new switch `order-by-last-updated-date`.
- Maps user input to `PackageOrder.LastPublished`.
- Integrates naturally into existing sort logic of `list`.

vexx32 reviewed (changes requested):
Looks pretty solid, nice self-contained change, just a couple minor comments.

You will need to rebase this branch against `develop` when you're done, just so the merge history is a bit tidier.

swaran18999 commented on src/chocolatey.resources/helpers/ChocolateyTabExpansion.ps1:
```diff
@@ -49,7 +49,7 @@ $commandOptions = @{
     install   = "--allow-downgrade --allow-empty-checksums --allow-empty-checksums-secure --apply-args-to-dependencies --apply-package-parameters-to-dependencies --cert='' --certpassword='' --disable-repository-optimizations --download-checksum='' --download-checksum-x64='' --download-checksum-type='' --download-checksum-type-x64='' --exit-when-reboot-detected --force-dependencies --forcex86 --ig ...
     license   = ""
-    list      = "--by-id-only --by-tag-only --detail --exact --id-only --id-starts-with --ignore-pinned --include-programs --page='' --page-size='' --prerelease --source='' --version=''"
+    list      = "--by-id-only --by-tag-only --detail --exact --id-only --id-starts-with --ignore-pinned --include-programs --page='' --page-size='' --prerelease --order-by-last-updated-date --show-last-updated-date --source='' --version=''"
```
change has been made

swaran18999 committed:
(chocolatey#1276) Add tab completion support for new list options
Update `ChocolateyTabExpansion.ps1` to include the new
`--show-last-updated-date` and `--order-by-last-updated-date` switches
in the `list` command's tab completion options.

swaran18999 committed:
(chocolatey#1276) Add tab completion support for new list options
Update `ChocolateyTabExpansion.ps1` to include the new
`--show-last-updated-date` and `--order-by-last-updated-date` switches
in the `list` command's tab completion options.

swaran18999 committed:
(chocolatey#1276) Refactor space changes and added last update dated

swaran18999 committed:
(chocolatey#1276) Refactor space changes and added last update dated

swaran18999 committed:
Merge branch 'list-command-with-date' of https://github.com/swaran18999/choco into list-command-with-date

swaran18999 committed:
Merge branch 'chocolatey:develop' into list-command-with-date

swaran18999 committed:
Merge branch 'list-command-with-date' of https://github.com/swaran18999/choco into list-command-with-date

swaran18999 committed:
(#1276) Persist .lastUpdated in package info

Add support for reading and writing the .lastUpdated file for package
information to track when packages were last modified.

Add LastUpdated property to ChocolateyPackageInformation domain to
store the timestamp. Update ChocolateyPackageInformationService.Get
to read .lastUpdated files.

Modify ChocolateyPackageInformationService.Save to persist or remove
.lastUpdated files, fetching the current timestamp from NugetService
via DateTime.Now.

swaran18999 committed:
(#1276]) Add LastUpdated ordering option for list command

Add support for ordering packages by last updated date in the list
command, providing users with a new sorting option to view packages
by their most recent updates.

swaran18999 committed:
(#1276]) Add LastUpdatedDate ordering with .lastUpdated file-based sorting

Add LastUpdatedDate ordering option that reads from .lastUpdated files
to sort packages by their actual update timestamps rather than remote
publication dates.

Add LastUpdatedDate enum value to PackageOrder with documentation
explaining it uses .lastUpdated file data. Update ChocolateyListCommand
to map --order-by-last-updated-date flag to PackageOrder.LastUpdatedDate
instead of LastUpdated.

Implement sorting logic in NugetList.ApplyPackageSort to read timestamp
from .lastUpdated files using the FaultTolerance pattern for error
handling. Sort packages by these timestamps in descending order with
secondary sorting by ID and version.

Make ChocolateyPackageInformationService.GetStorePath public to enable
NugetList to access package storage paths for reading .lastUpdated
files.

swaran18999 committed:
(#1276]) Consolidate LastUpdated ordering implementation

Remove duplicate LastUpdatedDate enum value and consolidate into the
existing LastUpdated ordering option. Update ChocolateyListCommand to
use PackageOrder.LastUpdated for the --order-by-last-updated-date flag.

Remove LastUpdatedDate case from NugetList.ApplyPackageSort switch
statement, leaving only the LastUpdated implementation that
reads from .lastUpdated files. Remove LastUpdated property from
integration test OData schema expectations as it is not needed.

swaran18999 committed:
(#1276]) Consolidate LastUpdated ordering implementation

Remove duplicate LastUpdatedDate enum value and consolidate into the
existing LastUpdated ordering option. Update ChocolateyListCommand to
use PackageOrder.LastUpdated for the --order-by-last-updated-date flag.

Remove LastUpdatedDate case from NugetList.ApplyPackageSort switch
statement, leaving only the LastUpdated implementation that
reads from .lastUpdated files. Remove LastUpdated property from
integration test OData schema expectations as it is not needed.

swaran18999 committed:
Merge branch 'develop' into list-command-with-date

vexx32 commented on src/chocolatey/infrastructure.app/services/ChocolateyPackageInformationService.cs:
```diff
@@ -353,7 +385,7 @@ public void Remove(IPackageMetadata package)
         }
 
-        private static string GetStorePath(IFileSystem fileSystem, string id, NuGetVersion version)
+        public static string GetStorePath(IFileSystem fileSystem, string id, NuGetVersion version)
```
I haven't found where this is used yet, but as a general rule we are *very* cautious making any APIs `public`. If this can be `internal` or something else non-public, that would be preferred. Given it's a static method, it also might make more sense to move it to a more easily-accessible place if we're needing to use it in multiple places.

vexx32 commented on src/chocolatey/infrastructure.app/services/NugetService.cs:
```diff
@@ -240,7 +241,7 @@ it is possible that incomplete package lists are returned from a command
                     packageInfo = _packageInfoService.Get(packageLocalMetadata);
                     deploymentlocation = packageInfo.DeploymentLocation;
                     sourceInstalledFrom = packageInfo.SourceInstalledFrom;
-
+                    lastUpdated = packageInfo.LastUpdated;
                     if (config.ListCommand.IncludeVersionOverrides)
```
```suggestion
                    lastUpdated = packageInfo.LastUpdated;

                    if (config.ListCommand.IncludeVersionOverrides)
```

vexx32 reviewed (changes requested):
Just about there in terms of the overall code, although we probably want to avoid making that method `public` if we can.

We'll also need to look at what tests can be added here to validate the behaviour automatically. If you'd like some pointers on that, there are some existing tests in `tests/chocolatey-tests/choco-list.Tests.ps1` that may be useful to reference, and we can discuss further as needed as well.

swaran18999 committed:
Merge branch 'list-command-with-date' of https://github.com/swaran18999/choco into list-command-with-date
